### PR TITLE
Add max column check when building vectorized expression

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

When giving column offset is invalid, the evaluation will panic. So this adds checks to the max column when building the RPN expression, so that invalid expressions will be detected early and return correct error instead of panic.

Extracted from #3898

- [x] Merge #4480 first

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

### CODE REVIEW

File: src/coprocessor/dag/rpn_expr/types/expr.rs [hunk 1]
```diff
@@ -106,4 +106,17 @@ impl AsRef<[RpnExpressionNode]> for RpnExpression {
     }
 }
 
+impl RpnExpression {
+    /// Gets the field type of the return value.
+    pub fn ret_field_type<'a>(&'a self, schema: &'a [FieldType]) -> &'a FieldType {
```
Comment (rleungx) [inline src/coprocessor/dag/rpn_expr/types/expr.rs]: Is this function related to this PR?

Comment (breeswish) [reply src/coprocessor/dag/rpn_expr/types/expr.rs]: Nop, but it is also needed for aggregation function framework. Such small changes are too many that it is not very friendly to create a PR for each of them.

Comment (AndreMouche) [reply src/coprocessor/dag/rpn_expr/types/expr.rs]: I do not think so, I think we should always focus on one thing or one group things in one PR. It seems this function is not related to this PR's title and may confuse the reviewers which may not understand the original issue

Comment (breeswish) [reply src/coprocessor/dag/rpn_expr/types/expr.rs]: Ok that's fine

File: src/coprocessor/dag/rpn_expr/types/expr_builder.rs [hunk 2]
```diff
@@ -119,12 +137,23 @@ fn get_eval_type(tree_node: &Expr) -> Result<EvalType> {
...
+    tree_node: Expr,
+    rpn_nodes: &mut Vec<RpnExpressionNode>,
+    max_columns: usize,
+) -> Result<()> {
     let offset = number::decode_i64(&mut tree_node.get_val()).map_err(|_| {
         Error::Other(box_err!(
             "Unable to decode column reference offset from the request"
         ))
     })? as usize;
+    if offset >= max_columns {
```
Comment (AndreMouche) [inline src/coprocessor/dag/rpn_expr/types/expr_builder.rs]: any test cover it?

Comment (breeswish) [reply src/coprocessor/dag/rpn_expr/types/expr_builder.rs]: This is a simple `if a >= b` check. I think it is not very necessary to verify that this line is correct or not.
